Support for LLVM 11

When i tried to run 'ninja check', i get the following error:

/opt/symcc/build# ninja check
[5/14] Performing configure step for 'SymRuntime'
-- The C compiler identification is GNU 10.0.1
-- The CXX compiler identification is GNU 10.0.1
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /opt/symcc/build/SymRuntime-prefix/src/SymRuntime-build
[8/14] Building CXX object CMakeFiles/Symbolize.dir/compiler/Symbolizer.cpp.o
FAILED: CMakeFiles/Symbolize.dir/compiler/Symbolizer.cpp.o 
/usr/bin/c++  -DSymbolize_EXPORTS -isystem /usr/lib/llvm-11/include -std=c++17 -Wredundant-decls -Wcast-align -Wmissing-include-dirs -Wswitch-default -Wextra -Wall -Winvalid-pch -Wredundant-decls -Wformat=2 -Wmissing-format-attribute -Wformat-nonliteral -Werror -fPIC   -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -MD -MT CMakeFiles/Symbolize.dir/compiler/Symbolizer.cpp.o -MF CMakeFiles/Symbolize.dir/compiler/Symbolizer.cpp.o.d -o CMakeFiles/Symbolize.dir/compiler/Symbolizer.cpp.o -c ../compiler/Symbolizer.cpp
../compiler/Symbolizer.cpp: In member function 'void Symbolizer::symbolizeFunctionArguments(llvm::Function&)':
../compiler/Symbolizer.cpp:38:77: error: no matching function for call to 'llvm::IRBuilder<>::CreateCall(llvm::Value* const&, llvm::ConstantInt*)'
   38 |                                                  IRB.getInt8(arg.getArgNo()));
      |                                              

I have libllvm11, llvm-11-tools and llvm-11-dev installed, but it doesn't address the compilation issue.
Any suggestions please?

--- a/compiler/Runtime.cpp
+++ b/compiler/Runtime.cpp
@@ -23,9 +23,9 @@ using namespace llvm;
 namespace {
 
 template <typename... ArgsTy>
-llvm::Value *import(llvm::Module &M, llvm::StringRef name, llvm::Type *ret,
+SymFnT import(llvm::Module &M, llvm::StringRef name, llvm::Type *ret,
                     ArgsTy... args) {
-#if LLVM_VERSION_MAJOR >= 9
+#if LLVM_VERSION_MAJOR >= 9 && LLVM_VERSION_MAJOR < 11
   return M.getOrInsertFunction(name, ret, args...).getCallee();
 #else
   return M.getOrInsertFunction(name, ret, args...);

--- a/compiler/Runtime.h
+++ b/compiler/Runtime.h
@@ -18,55 +18,61 @@
 #include <llvm/IR/InstrTypes.h>
 #include <llvm/IR/Module.h>
 
+#if LLVM_VERSION_MAJOR >= 9 && LLVM_VERSION_MAJOR < 11
+  using SymFnT = llvm::Value *;
+#else
+  using SymFnT = llvm::FunctionCallee;
+#endif
+
 /// Runtime functions
 struct Runtime {
   Runtime(llvm::Module &M);
 
-  llvm::Value *buildInteger{};
-  llvm::Value *buildInteger128{};
-  llvm::Value *buildFloat{};
-  llvm::Value *buildNullPointer{};
-  llvm::Value *buildTrue{};
-  llvm::Value *buildFalse{};
-  llvm::Value *buildBool{};
-  llvm::Value *buildSExt{};
-  llvm::Value *buildZExt{};
-  llvm::Value *buildTrunc{};
-  llvm::Value *buildBswap{};
-  llvm::Value *buildIntToFloat{};
-  llvm::Value *buildFloatToFloat{};
-  llvm::Value *buildBitsToFloat{};
-  llvm::Value *buildFloatToBits{};
-  llvm::Value *buildFloatToSignedInt{};
-  llvm::Value *buildFloatToUnsignedInt{};
-  llvm::Value *buildFloatAbs{};
-  llvm::Value *buildBoolAnd{};
-  llvm::Value *buildBoolOr{};
-  llvm::Value *buildBoolXor{};
-  llvm::Value *buildBoolToBits{};
-  llvm::Value *pushPathConstraint{};
-  llvm::Value *getParameterExpression{};
-  llvm::Value *setParameterExpression{};
-  llvm::Value *setReturnExpression{};
-  llvm::Value *getReturnExpression{};
-  llvm::Value *memcpy{};
-  llvm::Value *memset{};
-  llvm::Value *memmove{};
-  llvm::Value *readMemory{};
-  llvm::Value *writeMemory{};
-  llvm::Value *buildExtract{};
-  llvm::Value *notifyCall{};
-  llvm::Value *notifyRet{};
-  llvm::Value *notifyBasicBlock{};
+  SymFnT buildInteger{};
+  SymFnT buildInteger128{};
+  SymFnT buildFloat{};
+  SymFnT buildNullPointer{};
+  SymFnT buildTrue{};
+  SymFnT buildFalse{};
+  SymFnT buildBool{};
+  SymFnT buildSExt{};
+  SymFnT buildZExt{};
+  SymFnT buildTrunc{};
+  SymFnT buildBswap{};
+  SymFnT buildIntToFloat{};
+  SymFnT buildFloatToFloat{};
+  SymFnT buildBitsToFloat{};
+  SymFnT buildFloatToBits{};
+  SymFnT buildFloatToSignedInt{};
+  SymFnT buildFloatToUnsignedInt{};
+  SymFnT buildFloatAbs{};
+  SymFnT buildBoolAnd{};
+  SymFnT buildBoolOr{};
+  SymFnT buildBoolXor{};
+  SymFnT buildBoolToBits{};
+  SymFnT pushPathConstraint{};
+  SymFnT getParameterExpression{};
+  SymFnT setParameterExpression{};
+  SymFnT setReturnExpression{};
+  SymFnT getReturnExpression{};
+  SymFnT memcpy{};
+  SymFnT memset{};
+  SymFnT memmove{};
+  SymFnT readMemory{};
+  SymFnT writeMemory{};
+  SymFnT buildExtract{};
+  SymFnT notifyCall{};
+  SymFnT notifyRet{};
+  SymFnT notifyBasicBlock{};
 
   /// Mapping from icmp predicates to the functions that build the corresponding
   /// symbolic expressions.
-  std::array<llvm::Value *, llvm::CmpInst::BAD_ICMP_PREDICATE>
+  std::array<SymFnT, llvm::CmpInst::BAD_ICMP_PREDICATE>
       comparisonHandlers{};
 
   /// Mapping from binary operators to the functions that build the
   /// corresponding symbolic expressions.
-  std::array<llvm::Value *, llvm::Instruction::BinaryOpsEnd>
+  std::array<SymFnT, llvm::Instruction::BinaryOpsEnd>
       binaryOperatorHandlers{};
 };
 

--- a/compiler/Symbolizer.cpp
+++ b/compiler/Symbolizer.cpp
@@ -313,7 +313,7 @@ void Symbolizer::handleFunctionCall(CallBase &I, Instruction *returnPoint) {
   IRB.CreateCall(runtime.notifyCall, getTargetPreferredInt(&I));
 
   if (callee == nullptr)
-    tryAlternative(IRB, I.getCalledValue());
+    tryAlternative(IRB, I.getCalledOperand());
 
   for (Use &arg : I.args())
     IRB.CreateCall(runtime.setParameterExpression,
@@ -340,7 +340,7 @@ void Symbolizer::visitBinaryOperator(BinaryOperator &I) {
   // Binary operators propagate into the symbolic expression.
 
   IRBuilder<> IRB(&I);
-  Value *handler = runtime.binaryOperatorHandlers.at(I.getOpcode());
+  SymFnT handler = runtime.binaryOperatorHandlers.at(I.getOpcode());
 
   // Special case: the run-time library distinguishes between "and" and "or"
   // on Boolean values and bit vectors.
@@ -386,7 +386,7 @@ void Symbolizer::visitCmpInst(CmpInst &I) {
   // simply include either in the resulting expression.
 
   IRBuilder<> IRB(&I);
-  Value *handler = runtime.comparisonHandlers.at(I.getPredicate());
+  SymFnT handler = runtime.comparisonHandlers.at(I.getPredicate());
   assert(handler && "Unable to handle icmp/fcmp variant");
   auto runtimeCall =
       buildRuntimeCall(IRB, handler, {I.getOperand(0), I.getOperand(1)});
@@ -706,7 +706,7 @@ void Symbolizer::visitCastInst(CastInst &I) {
          {IRB.getInt8(I.getDestTy()->getIntegerBitWidth()), false}});
     registerSymbolicComputation(boolToBitConversion, &I);
   } else {
-    Value *target;
+    SymFnT target;
 
     switch (I.getOpcode()) {
     case Instruction::SExt:
@@ -888,7 +888,7 @@ CallInst *Symbolizer::createValueExpression(Value *V, IRBuilder<> &IRB) {
 }
 
 Symbolizer::SymbolicComputation
-Symbolizer::forceBuildRuntimeCall(IRBuilder<> &IRB, Value *function,
+Symbolizer::forceBuildRuntimeCall(IRBuilder<> &IRB, SymFnT function,
                                   ArrayRef<std::pair<Value *, bool>> args) {
   std::vector<Value *> functionArgs;
   for (const auto &[arg, symbolic] : args) {

--- a/compiler/Symbolizer.h
+++ b/compiler/Symbolizer.h
@@ -213,7 +213,7 @@ private:
 
   /// Like buildRuntimeCall, but the call is always generated.
   SymbolicComputation
-  forceBuildRuntimeCall(llvm::IRBuilder<> &IRB, llvm::Value *function,
+  forceBuildRuntimeCall(llvm::IRBuilder<> &IRB, SymFnT function,
                         llvm::ArrayRef<std::pair<llvm::Value *, bool>> args);
 
   /// Create a call to the specified function in the run-time library.
@@ -226,7 +226,7 @@ private:
   /// concrete (e.g., because they are compile-time constants), no call
   /// instruction is emitted and the function returns null.
   std::optional<SymbolicComputation>
-  buildRuntimeCall(llvm::IRBuilder<> &IRB, llvm::Value *function,
+  buildRuntimeCall(llvm::IRBuilder<> &IRB, SymFnT function,
                    llvm::ArrayRef<std::pair<llvm::Value *, bool>> args) {
     if (std::all_of(args.begin(), args.end(),
                     [this](std::pair<llvm::Value *, bool> arg) {
@@ -240,7 +240,7 @@ private:
 
   /// Convenience overload that treats all arguments as symbolic.
   std::optional<SymbolicComputation>
-  buildRuntimeCall(llvm::IRBuilder<> &IRB, llvm::Value *function,
+  buildRuntimeCall(llvm::IRBuilder<> &IRB, SymFnT function,
                    llvm::ArrayRef<llvm::Value *> symbolicArgs) {
     std::vector<std::pair<llvm::Value *, bool>> args;
     for (const auto &arg : symbolicArgs) {

